proposals/xxxx-package-manager-manifest-api-redesign.md

```diff
@@ -0,0 +1,674 @@
+
+    -    func predecessor() -> Version
+    }
+    ```
+    </p></details>
+
+* Make all properties of `Package` and `Target` mutable.
+
+    Currently, `Package` has three immutable and four mutable properties, and
+    `Target` has one immutable and one mutable property. We propose to make all
+    properties mutable to allow complex customization on the package object
+    after initial declaration.
+
+    <details>
+      <summary>View diff and example</summary>
+      <p>
+
+      Diff:
+    ```diff
+    final class Target {
+    -    let name: String
+    +    var name: String
+    }
+
+    final class Package {
+    -    let name: String
+    +    var name: String
+
+    -    let pkgConfig: String?
+    +    var pkgConfig: String?
+
+    -    let providers: [SystemPackageProvider]?
+    +    var providers: [SystemPackageProvider]?
+    }
+    ```
+
+    Example:
+    ```swift
+    let package = Package(
+        name: "FooPackage",
+        targets: [
+            Target(name: "Foo", dependencies: ["Bar"]),
+        ]
+    )
+
+    #if os(Linux)
+    package.targets[0].dependencies = ["BarLinux"]
+    #endif
+    ```
+    </p></details>
+
+* Change `Target.Dependency` enum cases to lowerCamelCase.
+
+    According to API design guidelines, everything other than types should be in lowerCamelCase.
+
+    <details>
+      <summary>View diff and example</summary>
+      <p>
+
+     Diff:
+    ```diff
+    enum Dependency {
+    -    case Target(name: String)
+    +    case target(name: String)
+
+    -    case Product(name: String, package: String?)
+    +    case product(name: String, package: String?)
+
+    -    case ByName(name: String)
+    +    case byName(name: String)
+    }
+    ```
+
+    Example:
+    ```diff
+    let package = Package(
+        name: "FooPackage",
+        targets: [
+            Target(
+                name: "Foo", 
+                dependencies: [
+    -                .Target(name: "Bar"),
+    +                .target(name: "Bar"),
+
+    -                .Product(name: "SwiftyJSON", package: "SwiftyJSON"),
+    +                .product(name: "SwiftyJSON", package: "SwiftyJSON"),
+                ]
+            ),
+        ]
+    )
+    ```
+    </p></details>
+
+* Add default parameter to the enum case `Target.Dependency.product`.
+
+    The associated value `package` in the (enum) case `product`, is an optional
+    `String`. It should have the default value `nil` so clients don't need to
+    write it if they prefer using explicit enum cases but don't want to specify
+    the package name i.e. it should be possible to write `.product(name:
+    "Foo")` instead of `.product(name: "Foo", package: nil)`.
+
+    If
+    [SE-0155](https://github.com/apple/swift-evolution/blob/master/proposals/0155-normalize-enum-case-representation.md)
+    is accepted, we can directly add a default value. Otherwise, we will use a
+    static factory method to provide default value for `package`.
```

Comment:
Note using static factory methods requires us to use a different name for the enum case currently. We should perhaps preemptively drop all direct uses of the enum initializer, in the interest of ensuring future extensibility.

I noticed this when making the build setting API proposal... we will need a way to spell another Target.Dependency.